[lldb][lldb-dap] Implement jump to cursor

da-viper opened the pull request:
Fixes #109335

Depends on #130435

da-viper committed:
[lldb-dap] implement jump to cursor.

da-viper committed:
[lldb][lldb-dap] add jump to cursor tests

da-viper committed:
[lldb-dap] Rever removing libXML version

da-viper committed:
[lldb][lldb-dap] fix code format

vogelsgesang commented on lldb/tools/lldb-dap/CMakeLists.txt:
```diff
@@ -50,6 +50,8 @@ add_lldb_tool(lldb-dap
   Handler/DisconnectRequestHandler.cpp
   Handler/EvaluateRequestHandler.cpp
   Handler/ExceptionInfoRequestHandler.cpp
+        Handler/GoToRequestHandler.cpp
+        Handler/GoToTargetsRequestHandler.cpp
```
inconsistent indentation

vogelsgesang commented on lldb/test/API/tools/lldb-dap/gotoTarget/main.c:
```diff
@@ -0,0 +1,11 @@
+  int var_2 = 40; // goto 1
+
+  return 0;
+}
```
missing trailing new line

vogelsgesang commented on lldb/test/API/tools/lldb-dap/gotoTarget/TestDAP_gotoTarget.py:
```diff
@@ -0,0 +1,59 @@
+            response["success"], True, "expects success to go to a target id"
+        )
+
+        var_1_object = self.dap_server.get_local_variable("var_1")
+        self.assertEqual(first_var_1_object["value"], var_1_object["value"])
```
Currently, this test case does not really verify where code execution continues after the go-to. It only verifies that `var1` stays unchanged, but it does not check that `var_2 = 40` is indeed executed next.

Can we write this test case to check for `var_1 = 10, var_2 = 40`, i.e. a state which should not normally be reachable, except if the goto skipped over line 6?

vogelsgesang commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -79,6 +79,27 @@ enum class PacketStatus {
+  /// \return id assigned to this line_entry.
+  uint64_t InsertLineEntry(lldb::SBLineEntry line_entry);
+
+  /// clears all line entries and reset the generated ids.
```
```suggestion
  /// Clears all line entries and reset the generated ids.
```

vogelsgesang commented on lldb/tools/lldb-dap/Handler/GoToRequestHandler.cpp:
```diff
@@ -0,0 +1,103 @@
+//     continue to run.\nThis makes it possible to skip the execution of code or
+//     to execute code again.\nThe code between the current location and the
+//     goto target is not executed but skipped.\nThe debug adapter first sends
+//     the response and then a `stopped` event with reason `goto`.\nClients
```
> `stopped` event with reason `goto`

Are we doing this correctly? Please also check for it in the test case

da-viper replied:
Created an new helper function as lldb does not have a `eStopReason` for `goto` and it also does not send an event when using jump from the command line.

vogelsgesang commented on lldb/tools/lldb-dap/Handler/GoToRequestHandler.cpp:
```diff
@@ -0,0 +1,103 @@
+//     goto target is not executed but skipped.\nThe debug adapter first sends
+//     the response and then a `stopped` event with reason `goto`.\nClients
+//     should only call this request if the corresponding capability
+//     `supportsGotoTargetsRequest` is true (because only then goto targets
+//     exist that can be passed as arguments).", "properties": {
+//       "command": {
```
```suggestion
//     `supportsGotoTargetsRequest` is true (because only then goto targets
//     exist that can be passed as arguments).",
//.    "properties": {
//       "command": {
```

vogelsgesang commented on lldb/tools/lldb-dap/Handler/GoToRequestHandler.cpp:
```diff
@@ -0,0 +1,103 @@
+  SendThreadStoppedEvent(dap);
+}
+
+} // namespace lldb_dap
```
missing new line

vogelsgesang commented on lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp:
```diff
@@ -0,0 +1,120 @@
+//===-- GoToTargetsRequestHandler.cpp
+//--------------------------------------===//
```
incorrect linebreak in comment

da-viper committed:
Update lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp

Co-authored-by: Adrian Vogelsgesang <adrian.vogelsgesang@tum.de>

da-viper committed:
Update lldb/tools/lldb-dap/DAP.h

Co-authored-by: Adrian Vogelsgesang <adrian.vogelsgesang@tum.de>

da-viper commented on lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp:
```diff
@@ -0,0 +1,120 @@
+  const auto goto_column =
+      GetInteger<uint64_t>(arguments, "column").value_or(0u);
+
+  lldb::SBLineEntry line_entry{};
+  const lldb::SBFileSpec file_spec(path.c_str(), true);
+  line_entry.SetFileSpec(file_spec);
+  line_entry.SetLine(goto_line);
+  line_entry.SetColumn(goto_column);
```
[SBThread](https://lldb.llvm.org/python_api/lldb.SBThread.html#lldb.SBThread.JumpToLine) currently only provides [jump to line](https://lldb.llvm.org/python_api/lldb.SBThread.html#lldb.SBThread.JumpToLine) and no way to check if you can jump to a line.  If the current line is empty it correctly jump to the next line with statements. 

I wanted to try validating it the line_entry with `SBLineEntyry::IsValid` but we cannot guarantee if the previous / next line is in the current scope or function.

walter-erquinigo replied:
Somehow I think we shouldn't support jump based on columns. You might find yourself breaking the semantics of the code.

da-viper replied:
@walter-erquinigo  @vogelsgesang 

I was thinking instead we only support jumps within a function for the same reason as above. 

it would be something like this 

```py
def is_line_in_func(fn, line) ->:
       s_addr = fn.addr
       e_addr = fn.end_addr 
       
       return (s_addr <= line.addr )and (line.addr <= e_addr)

def get_line_valid_entry(file_spec, line, target):

```

da-viper replied:
I used breakpoints instead because they are able to resolve if the line location is valid. then use it to create targets for the goto request

vogelsgesang replied:
> Somehow I think we shouldn't support jump based on columns. You might find yourself breaking the semantics of the code.

I think in an ideal world we would allow "go to" for all line entries marked as `is_stmt`. Currenlty, `is_stmt` is set also based on line numbers. There are also some discussions on improving the `is_stmt` on Discourse, though https://discourse.llvm.org/t/rfc-improving-is-stmt-placement-for-better-interactive-debugging/82668/13.

CC @labath, since you were also active on that discourse thread

da-viper committed:
Update lldb/tools/lldb-dap/Handler/GoToRequestHandler.cpp

Co-authored-by: Adrian Vogelsgesang <adrian.vogelsgesang@tum.de>

da-viper committed:
[lldb][lldb-dap] add review commits

da-viper committed:
[lldb][lldb-dap] add review changes

da-viper committed:
[lldb][lldb-dap] Update jump to cursor test

da-viper committed:
[lldb][lldb-dap] add new test for goto execute again.

da-viper committed:
[lldb][lldb-dap] add helper function `SendThreadGotoEvent`

LLDB currently does not have a `StopReason` for goto so I cannot use `SendThreadStoppedEvent` as I need to put the goto reason.

Signed-off-by: Ebuka Ezike <yerimyah1@gmail.com>

da-viper committed:
[lldb][lldb-dap] Rename `gotoTarget` to `gotoTargets` and improve validation.

Renamed test files, directories, and references from `gotoTarget` to `gotoTargets` for consistency. Added validation to confirm stopped events include a "goto" reason after goto requests in the test suite.

Signed-off-by: Ebuka Ezike <yerimyah1@gmail.com>

walter-erquinigo commented on lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp:
```diff
@@ -0,0 +1,121 @@
+    const llvm::json::Object &request) const {
+  llvm::json::Object response;
+  FillResponse(request, response);
+  const auto *arguments = request.getObject("arguments");
+  const auto *source = arguments->getObject("source");
+  const std::string path = GetString(source, "path").str();
+
+  const auto goto_line = GetInteger<uint64_t>(arguments, "line").value_or(0u);
+  const auto goto_column =
```
try not to use auto in new code unless the type is horrible

walter-erquinigo commented on lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp:
```diff
@@ -0,0 +1,121 @@
+  line_entry.SetLine(goto_line);
+  line_entry.SetColumn(goto_column);
+
+  const auto target_id = dap.goto_id_map.InsertLineEntry(line_entry);
+  llvm::json::Array response_targets;
+  const auto target_line = line_entry.GetLine();
+  const auto target_column = line_entry.GetColumn();
```
no auto

da-viper replied:
> Somehow I think we shouldn't support jump based on columns. You might find yourself breaking the semantics of the code.

Makes sense since `SBThread::JumpToLine` takes a filespec and line number

da-viper committed:
Update lldb/test/API/tools/lldb-dap/gotoTargets/TestDAP_gotoTargets.py

Co-authored-by: Adrian Vogelsgesang <adrian.vogelsgesang@tum.de>

da-viper committed:
Update lldb/test/API/tools/lldb-dap/gotoTargets/TestDAP_gotoTargets.py

Co-authored-by: Adrian Vogelsgesang <adrian.vogelsgesang@tum.de>

da-viper committed:
Update lldb/test/API/tools/lldb-dap/gotoTargets/TestDAP_gotoTargets.py

Co-authored-by: Adrian Vogelsgesang <adrian.vogelsgesang@tum.de>

da-viper committed:
Update lldb/test/API/tools/lldb-dap/gotoTargets/TestDAP_gotoTargets.py

Co-authored-by: Adrian Vogelsgesang <adrian.vogelsgesang@tum.de>

da-viper committed:
[lldb][lldb-dap] use breakpoints to verify goto location.

da-viper committed:
Add Release notes

JDevlieghere commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -79,6 +79,27 @@ enum class PacketStatus {
+  void Clear();
+
+private:
+  uint64_t NewSpecId();
```
```suggestion
  uint64_t NewSpecID();
```

JDevlieghere commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -79,6 +79,27 @@ enum class PacketStatus {
+  uint64_t NewSpecId();
+
+  llvm::DenseMap<uint64_t, lldb::SBLineEntry> line_entries;
+  uint64_t new_id = 0ul;
```
```suggestion
  uint64_t new_id = 0;
```

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -899,6 +899,27 @@ lldb::SBError DAP::WaitForProcessToStop(uint32_t seconds) {
+}
+
+void Gotos::Clear() {
+  new_id = 0UL;
```
```suggestion
  new_id = 0;
```

JDevlieghere commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -79,6 +79,27 @@ enum class PacketStatus {
+  /// \return id assigned to this line_entry.
+  uint64_t InsertLineEntry(lldb::SBLineEntry line_entry);
+
+  /// Clears all line entries and reset the generated ids.
+  void Clear();
```
It doesn't look this is actually used?

da-viper replied:
Yes, It is used here. 

https://github.com/llvm/llvm-project/blob/a2a2cb85e5f17a7468970c6014947918d9805e80/lldb/tools/lldb-dap/DAP.h#L378-L379

JDevlieghere commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -209,6 +230,7 @@ struct DAP {
   // empty; if the previous expression was a variable expression, this string
   // will contain that expression.
   std::string last_nonempty_var_expression;
+  Gotos goto_id_map;
```
This is more than just a map, right?
```suggestion
  Gotos gotos;
```

JDevlieghere commented on lldb/tools/lldb-dap/Handler/GoToRequestHandler.cpp:
```diff
@@ -0,0 +1,118 @@
+
+namespace lldb_dap {
+
+/// Creates an \p StoppedEvent with the reason \a goto
```
```suggestion
/// Creates an \p StoppedEvent with the reason \a goto.
```

JDevlieghere commented on lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp:
```diff
@@ -0,0 +1,165 @@
+
+static llvm::SmallVector<lldb::SBLineEntry>
+GetLineValidEntry(DAP &dap, const lldb::SBFileSpec &file_spec, uint32_t line) {
+  // disable breakpoint listeners so they do not send events to the DAP client.
```
```suggestion
  // Disable breakpoint listeners so they do not send events to the DAP client.
```

JDevlieghere commented on lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp:
```diff
@@ -0,0 +1,165 @@
+  constexpr auto event_mask = lldb::SBTarget::eBroadcastBitBreakpointChanged;
+  listener.StopListeningForEvents(broadcaster, event_mask);
+
+  // create a breakpoint to resolve the line if it is on an empty line.
```
```suggestion
  // Create a breakpoint to resolve the line if it is on an empty line.
```

JDevlieghere commented on lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp:
```diff
@@ -0,0 +1,165 @@
+  lldb::SBListener listener = dap.debugger.GetListener();
+  lldb::SBBroadcaster broadcaster = dap.target.GetBroadcaster();
+  constexpr auto event_mask = lldb::SBTarget::eBroadcastBitBreakpointChanged;
+  listener.StopListeningForEvents(broadcaster, event_mask);
```
This seems somewhat iffy. It's tacitly synchronizing with the even thread which is going to be hard to reason about if we ever have to debug a breakpoint event not getting broadcast. 

I wonder if we can avoid this complexity by using an internal breakpoint that doesn't have the dap label. I'm not sure if we broadcast breakpoint changed events for internal breakpoints (I would expected we don't but I didn't look at the code to confirm), but even if we do, we should ignore breakpoints that weren't created through lldb-dap (by checking the label).

da-viper replied:
At the moment we do not broadcast breakpoint that is not owned by dap. But I think we should as it will make setting breakpoint from the debug console possible.

walter-erquinigo replied:
yes, I think we shuold do that with broadcasting. The IDE should transparently behave well after cli commands.

I don't see clearly how broadcasting would simplify this code, but I trust Jonas. Having to stop listening for events for some time open the door to some hard-to-catch bugs and we should avoid that whenever possible.

clayborg replied:
Seems like we are using the breakpoint API just so we can answer the questions "tell me all of the locations that map to source file and line". I would prefer to expose the source file + line resolving code in a `SBTarget` API so we can avoid having to use breakpoints just so we can get all of the locations. The API I am thinking would be:
```
lldb::SBSymbolContextList lldb::SBTarget::ResolveContexts(lldb::SBFileSpec &file_spec, uint32_t line);
```
And this returns the same info as the breakpoint location contexts. LLDB breakpoint code uses address resolvers and search filters, and we can just use the same APIs as the breakpoints.

clayborg replied:
Using breakpoints and disabling notifications seems like a bad side affect of using the Breakpoint APIs to work around a missing and needed public API.

da-viper replied:
Is there like a procedure in adding new SB Public API ?

ashgti replied:
I think we can use the `lldb.target.modules[0].FindCompileUnits()` to find this information without using a breakpoint.

Trying this out on the stepInTargets binary:

```
$ lldb lldb-dap/stepInTargets/TestDAP_stepInTargets.test_basic/a.out
(lldb) script
>>>[line_entry for line_entry in lldb.target.modules[0].FindCompileUnits(lldb.SBFileSpec("lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp"))[0].compile_unit]
[
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:2, 
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:2:38,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:2:44,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:2:42,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:2:31,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:4:15,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:6:15,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:8,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:9:7,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:9:16,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:9:3,
  lldb/test/API/tools/lldb-dap/stepInTargets/main.cpp:10:3,
]
```

I'm not sure how we should scope those. We could use the block to scope the targets to the same block as the requested file:line:column.

clayborg replied:
This will work for code in the current source file when the source file itself matchs the name of a compile unit, but it won't work for inline functions. The line table above could easily contain entryes for "/.../stl/vector:123" where this is the `vector` header file from the STL. There will be no top level compile unit for this. 

So we really should add the new API so that it can use the same source file searching code that the breakpoints use.

It is easy to add things to the public API by following the API rules and making sure people agree on the signature and function location. Typically for things that want to search for things in a `lldb::SBModule`, we can add a `lldb::SBModule` API, but we might also want to add an API at the `lldb::SBTarget` so that it can search all modules so we might want the following APIs:

```
/// Find all locations for a file and line in this lldb::SBModule
lldb::SBSymbolContextList lldb::SBModule::ResolveContexts(lldb::SBFileSpec &file_spec, uint32_t line);
/// Find all locations for a file and line in all lldb::SBModule in the lldb::SBTarget
lldb::SBSymbolContextList lldb::SBTarget::ResolveContexts(lldb::SBFileSpec &file_spec, uint32_t line);
```

clayborg replied:
A patch that adds the above APIs could be done in a separate patch. It can easily to unit tested.

clayborg replied:
We might also want to add a parameter that specifies if we want to catch inline line entries, or just match line entries for compile units:

```
/// Find all locations for a file and line in this lldb::SBModule
lldb::SBSymbolContextList lldb::SBModule::ResolveContexts(lldb::SBFileSpec &file_spec, uint32_t line, bool find_inlines);
/// Find all locations for a file and line in all lldb::SBModule in the lldb::SBTarget
lldb::SBSymbolContextList lldb::SBTarget::ResolveContexts(lldb::SBFileSpec &file_spec, uint32_t line, bool find_inlines);
```

@JDevlieghere @jimingham let me know what you think of the above APIs? I think it would be very useful to expose the breakpoint location search features as an API.

jimingham replied:
This seems generally useful, and it is silly to have to actually set breakpoints to do this kind of symbol search.

I think at this point we should always allow column information when we do file & line type queries along with the file & line.  I almost wonder if we should have an SBLineEntrySpec to use for these sorts of queries so we don't have to repeat all the parameters.  But regardless, we should allow users to also pass in column info here.

In the case of breakpoints we will allow inexact matches if the user asks for them.  That's useful for setting breakpoints, but I'm not sure it's all that useful for a "ResolveContexts" type API.  So I don't think you need to add a `bool exact_match` but you should document the behavior since people might be surprised if it doesn't match what `break set` of the same file & line returns.

jimingham replied:
There are other types of queries that might be interesting here, like "file and line contained in another file or function" - when you're looking for inlining that's handy.  I'm not saying that we should support those up front, but if we can think of a couple more things like that that we'd want to do, it might be worth coming up with a container (like SBLineEntrySpec) that we can then add to as we go w/o having to change the API's.

vogelsgesang replied:
If we add a new API, we should make sure it also covers the "get column breakpoints" use case. See the [current source code](https://github.com/llvm/llvm-project/blob/b82fd7110972c52cf4e58bf08b65bce7a91ecb0e/lldb/tools/lldb-dap/Handler/BreakpointLocationsHandler.cpp#L157) which iterates over the line tables manually. In the PR which introduced column-breakpoint support for lldb-dap, we already had a [similar discussion on exposing a new SB-API](https://github.com/llvm/llvm-project/pull/113787/files/4e05a8e1bb01e4b1472e9ec81111d9a169cdf793#r1827507293).

For the API, this would mean that:
* We should be able to search for a line-range and not only for a single line
* We should be able to restrict the column numbers
* We need a way to return all column positions within the line(s) on which we could potentially stop

JDevlieghere commented on lldb/tools/lldb-dap/Handler/GoToRequestHandler.cpp:
```diff
@@ -0,0 +1,118 @@
+  llvm::json::Object response;
+  FillResponse(request, response);
+
+  auto SendError = [&](auto &&message) {
+    response["success"] = false;
+    response["message"] = message;
+    dap.SendJSON(llvm::json::Value(std::move(response)));
+  };
```
This is a neat little helper. Can we move this up into the RequestHandler class or the DAP class and use it in other handlers?

vogelsgesang replied:
IMO, we should not factor out this helper, because:

1. We already have a better abstraction.
2. This helper is encouraging to violate the debug adapter protocol.

**Better abstraction**: In #129155, @ashgti already introduced the `protocol::Response` class, which also provides the `message` header. In #130090, those messages got integrated with our request handlers. IMO, we should change this PR to use the new mechanism introduced in #130090, instead of using the legacy mechanism.

**Violation of the debug adapter protocol specification**: This helper sets the message to an arbitrary string, such as `"Arguments is empty"`. However, this is not the correct usage of DAP.

The `message` field of the [Response](https://microsoft.github.io/debug-adapter-protocol/specification#Base_Protocol_Response) should be a machine-readable enum (`cancelled`, `notStopped`, ...). It should not contain the human-readable the error message and will not be displayed to users.

Instead, the user-readable message should use the [ErrorResponse](https://microsoft.github.io/debug-adapter-protocol/specification#Base_Protocol_ErrorResponse), i.e. the user-readable message should be in `body. format`, and `body.showUser` should probably be set to true.

This is currently done incorrectly by lldb-dap everywhere. This is also, why users currently get no feedback if, e.g., setting a variable failed. The new mechanism introduced in #130090 still does not handle this correctly. But if we consistently adopt the new mechanism, we at least only need to fix a single place to become standard-compliant

(That being said, I would not block this PR on using `RequestHandler` instead of `LegacyRequestHandler,` in case @da-viper does not want to invest the time to update his code to the newer `RequestHandler` mechanism)

ashgti replied:
I had been meaning to follow up on this with a `DAPError` class, I sent #132255 to implement that.

Also, as vogelsgesang said, I wouldn't block this PR on using that once #132255 is in, that should help clean up the code a bit.

da-viper replied:
The new protocol currently does not implement  StoppedEvent. Not sure If I should implement it in this PR also. 

It may take longer since @clayborg  is proposing adding a new SB Public API function. 

Also the current new protocol does not take into account if we need to send an event after sending a response. Currently it will require repeating the  `operator()(const protocol::Request &request)` code for any request that needs to send an event after a response such as stepIn, StepOut and Next

JDevlieghere commented on lldb/tools/lldb-dap/Handler/GoToRequestHandler.cpp:
```diff
@@ -0,0 +1,118 @@
+  };
+
+  const auto *goto_arguments = request.getObject("arguments");
+  if (goto_arguments == nullptr) {
+    SendError("Arguments is empty");
+    return;
+  }
```
Since this function returns a void, and the lambda does too, you can do:
```suggestion
  if (goto_arguments == nullptr) 
    return SendError("Arguments is empty");
```

JDevlieghere commented on lldb/tools/lldb-dap/Handler/InitializeRequestHandler.cpp:
```diff
@@ -406,7 +406,7 @@ void InitializeRequestHandler::operator()(
   body.try_emplace("supportsRestartFrame", false);
   // The debug adapter supports the gotoTargetsRequest.
-  body.try_emplace("supportsGotoTargetsRequest", false);
+  body.try_emplace("supportsGotoTargetsRequest", true);
```
You'll need to rebase this on top of #131943 and move it into the request handler.

da-viper committed:
[lldb-dap] Update with review changes

vogelsgesang commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -113,6 +113,20 @@ class ExceptionInfoRequestHandler : public RequestHandler {
   void operator()(const llvm::json::Object &request) const override;
 };
 
+class GoToRequestHandler : public RequestHandler {
```
please rebase on `main`, and make sure to use the new, non-legacy mechanism for those new request handlers introduced in #130090

vogelsgesang replied:
Good point.

Currently, we are using the naming from the [Debug Adapter Protocol spec](https://github.com/microsoft/debug-adapter-protocol/blob/main/debugAdapterProtocol.json) directly (That's also where we copied all the comments from).

As such, we also copied over the camelCase style from the DAP spec. Not sure if we want to stick to the DAP style or LLDB style here. Afaik, this is not a discussion we had so far.

@ashgti @JDevlieghere WDYT?

da-viper replied:
There is an issue with the way to new protocol works, If you need to send and event after a response. (currently needed for goto, next , stepIn , stepOut) .  It does not handle that case. 
Instead you have to copy the implementation of `void operator()(const protocol::Request &request)` from the  `RequestHandler`

vogelsgesang commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -113,6 +113,20 @@ class ExceptionInfoRequestHandler : public RequestHandler {
+class GoToRequestHandler : public RequestHandler {
+public:
+  using RequestHandler::RequestHandler;
+  static llvm::StringLiteral getCommand() { return "goto"; }
```
After rebasing on main... This function got renamed in 1028ea9e2641341436555347fb65d6f2eafcfbdd

```suggestion
  static llvm::StringLiteral GetCommand() { return "goto"; }
```

da-viper committed:
[lldb-dap] update gototargets to use the new protocol

da-viper committed:
[lldb-dap] Fix code format

da-viper committed:
[lldb][lldb-dap] enable gotoTargets capability

da-viper committed:
[lldb][lldb-dap] use std::vector instead of llvm::DenseMap.

da-viper committed:
[lldb][lldb-dap] use llvm::smallVector as gotargets are usually few

da-viper committed:
[lldb][lldb-dap] Update with review changes

Remove disabling the breakpoint event listener.

Some of the error are not meant to be user facing as they cannot do anything with the message received.

da-viper commented on lldb/tools/lldb-dap/Handler/GoToTargetsRequestHandler.cpp:
```diff
@@ -0,0 +1,91 @@
+
+  // clean up;
+  dap.target.BreakpointDelete(goto_bp.GetID());
+  listener.StartListeningForEvents(broadcaster, event_mask);
```
Since the breakpoint is not valid it would not create a valid breakpoint that can be deleted
